### ISSUE TITLE
Approximate spherical lighting

### ISSUE THREAD

**Issue by rdococ** [minetest/minetest#8818]
#7887 was a suggestion for "sub-node lighting" and spherically-spreading lighting. Sub-node lighting is probably impossible for the foreseeable future, but you could approximate spherical lighting with a simple rule.

Light would be able to spread diagonally, but only at certain light levels.

In the example below, light can only spread diagonally above a certain light level. Below that light level, light would not be able to spread diagonally. Dimmer lights would appear diamond-shaped, like they do now, but brighter lights will spread out more like a rounded square.

Here is an example in 2D.

![](https://i.imgur.com/wA5zamX.png)

If you want all lights to spread out in similar shapes, instead of dimmer lights spreading out in the current diamond shape and brighter lights spreading out in more of a rounded square shape, then you could use a different rule, such as distributing the number of "light levels that can spread diagonally" evenly amongst the total number of light levels.

**Comment by ClobberXD** [comment]
I support this concept. But treating brighter lights differently from dimmer lights w/ respect to spread is something I don't think is right. IMO, light from all light sources should spread this way irrespective of brightness, maybe with variable spread radius or intensity based on the brightness instead.

**Comment by HybridDog** [comment]
What happens if you do it the other way round, i.e. spread diagonally for high light levels and all directions for low levels?

ClobberXD, I think that the brightness is already unrealistic and/or inconsistent, for example because of the sRGB colourspace.

**Comment by paramat** [comment]
I don't think it is at all desirable to approximate spherical lighting spread in any way, for obvious reasons. This is a cubic-voxel game engine where most things are cubic or square, and where things spherical or circular in reality are cubic or square, primarily for the purpose of abstraction for simplicity and performance. So diamond light spread, being a rotated square, is ideal, any non-square shape would be less suitable and look more out of place.

The result of your idea looks worse (it has distracting amounts of structure) than the current simple and consistent diamond spread.
Having behaviour of light spread alter depending on light level is inconsistent and would look bad.
Diagonal light spread requires more complex code and spread to 8 neighbours instead of 4, because light spread is recursive and already very intensive, diagonal spread would probably cause a big performance problem.

:-1:

**Comment by TumeniNodes** [comment]
I'm wondering what it would take just to possibly add some flickering to lighting (basically torches) which can also be avaialable to other mods to use, such as lanterns, etc.?

In my messed up mind, achieving flickering should be easier than adding spherical lighting and might be a good starting point for some lighting system changes?

(Hopefully this is not OT by much)
I would think flickering would be a very nice feature... (but would that be an on/off feature or constant?)

I have found it a bit odd that we cannot add a sepia tone to lighting (at least for light nodes) just to start

**Comment by Moth-Tolias** [comment]
do you mean flickering or modulation? flickering would be exceedingly irritating.

**Comment by TumeniNodes** [comment]
> do you mean flickering or modulation? flickering would be exceedingly irritating.

simulating a flame flickering
I using spatial modulation can produce such a random effect...yeh... then _that_ one
xD

**Comment by lhofhansl** [comment]
I'm a big fan of spherical light spreading. Yes, Minetest is a voxel game, and we'd still calculate light per voxel, but that does not mean that need to maintain that weird looking and unnatural square'ish light spread.

As for flickering, it should be easy to add this via shaders. But note that we cannot distinguish between different artificial light sources, only between sunlight and non-sunlight. So all non-sunlight would need to flicker... Unless we add something new here.

**Comment by Fixer-007** [comment]
Will be interesting to see how it looks in game, what's the looks, performance, etc

**Comment by numberZero** [comment]
> edit: ahhh setMaterialFlag(EMF_LIGHTING,true) has something to do with it.

That’s for pre-shader hardware lighting.

> Diagonal light spread requires more complex code and spread to 8 neighbours instead of 4, 

Check your math, a cube has 6 neighbors) plus 12 on edges and 8 on vertices. That’s a lot so requires a very efficient code. Without bit packing, it only slows things down (probably). (btw the current code uses BFS and not recursion, doesn’t it?)
